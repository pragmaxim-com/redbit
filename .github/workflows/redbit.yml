name: Update README with Example Code

on:
  push:
    branches:
      - master
  workflow_dispatch:

permissions:
  contents: write

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repository
        uses: actions/checkout@v4

      - name: Setup Rust
        uses: dtolnay/rust-toolchain@nightly

      - name: Cache Rust dependencies
        uses: Swatinem/rust-cache@v2
        with:
          shared-key: ${{ runner.os }}-ci
          cache-on-failure: true

      - name: Run tests
        run: |
          cargo test --workspace --features ci

  integration:
    needs: test
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repository
        uses: actions/checkout@v4

      - name: Setup Rust
        uses: dtolnay/rust-toolchain@nightly

      - name: Cache Rust dependencies
        uses: Swatinem/rust-cache@v2
        with:
          shared-key: ${{ runner.os }}-ci
          cache-on-failure: true

      - name: Run tests
        run: |
          cargo test --package demo --features integration,ci
          cargo test --package btc --features integration,ci
          cargo test --package cardano --features integration,ci
          cargo test --package ergo --features integration,ci

  criterion:
    needs: integration
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repository
        uses: actions/checkout@v4

      - name: setup | rust
        uses: dtolnay/rust-toolchain@nightly

      - name: Cache Rust dependencies
        uses: Swatinem/rust-cache@v2
        with:
          shared-key: ${{ runner.os }}-ci
          cache-on-failure: true

      - name: Run benchmarks
        run: |
          cargo bench --package demo --bench demo_benchmark -- --quiet
          cargo bench --package btc --bench btc_benchmark  -- --quiet
          cargo bench --package cardano --bench cardano_benchmark -- --quiet
          cargo bench --package ergo --bench ergo_benchmark  -- --quiet

      - name: Upload Criterion report to Pages artifact
        id: deployment
        uses: actions/upload-pages-artifact@v3
        with:
          path: target/criterion/

  bench:
    needs: criterion
    runs-on: ubuntu-latest
    steps:
      - name: Checkout Repository
        uses: actions/checkout@v4

      - name: setup | rust
        uses: dtolnay/rust-toolchain@nightly

      - name: Cache Rust dependencies
        uses: Swatinem/rust-cache@v2
        with:
          shared-key: ${{ runner.os }}-ci
          cache-on-failure: true

      - name: Run Built-in Rust Benchmarks and Extract Metrics
        run: |
          run_bench_and_print() {
            PACKAGE=$1
            OUT_FILE="${PACKAGE}_bench_output.txt"
            CSV_FILE="${PACKAGE}_benchmarks.csv"

            echo "Running benchmarks for $PACKAGE ..."
            cargo bench --package "$PACKAGE" --features bench -- --nocapture | tee "$OUT_FILE"

            # Extract and process lines with `bench:`
            grep 'bench:' "$OUT_FILE" | while read -r line; do
              # Extract full function name regardless of spacing
              FUNCTION=$(echo "$line" | sed -n 's/^test \([^ ]*\) \+\.\.\. bench:.*/\1/p')

              # Extract time in ns/iter
              TIME_NS=$(echo "$line" | grep -oP 'bench:\s+\K[0-9.,]+' | tr -d ',')

              # Convert TIME_NS to ops/sec
              OPS_PER_S=$(awk "BEGIN {if ($TIME_NS > 0) printf \"%.0f\", 1000000000 / $TIME_NS; else print 0}")

              echo "$FUNCTION,$OPS_PER_S" >> "$CSV_FILE"
            done

            sort -t, -k2 -n "$CSV_FILE" -o "$CSV_FILE"
          }

          run_bench_and_print demo
          cargo bench --package redbit --features bench

      - name: Update README with documentation
        run: |
          LIB_CONTENT=$(cat chains/demo/src/model_v1.rs | sed 's/^/    /')
          MAIN_CONTENT=$(cat chains/demo/src/showcase.rs | sed 's/^/    /')

          # Update BEGIN_LIB and BEGIN_MAIN sections
          awk -v lib_content="$LIB_CONTENT" -v main_content="$MAIN_CONTENT" '
            BEGIN { in_lib = 0; in_main = 0 }
            /<!-- BEGIN_LIB -->/ { print; print "```rust"; print lib_content; print "```"; in_lib = 1; next }
            /<!-- END_LIB -->/ { print; in_lib = 0; next }
            /<!-- BEGIN_MAIN -->/ { print; print "```rust"; print main_content; print "```"; in_main = 1; next }
            /<!-- END_MAIN -->/ { print; in_main = 0; next }
            !in_lib && !in_main { print }
          ' README.md > README.md.tmp && mv README.md.tmp README.md

      - name: Update Readme with bench results
        run: |
          # Prepare formatted benchmark lines
          HEADER=$(printf "%-60s %10s\n" "function" "ops/s")
          BODY=$(cat demo_benchmarks.csv | awk -F, '{ printf "%-60s %10s\n", $1, $2 }')
          FORMATTED_BENCH="$HEADER\n-------------------------------------------------------------\n$BODY"

          # Escape double quotes and backslashes
          ESCAPED_BENCH=$(echo -e "$FORMATTED_BENCH" | sed 's/\\/\\\\/g; s/"/\\"/g')

          # Replace section between BEGIN_BENCH and END_BENCH
          awk -v bench_block="$ESCAPED_BENCH" '
            BEGIN { in_bench = 0 }
            /<!-- BEGIN_BENCH -->/ {
              print;
              print "```";
              print bench_block;
              print "```";
              in_bench = 1;
              next
            }
            /<!-- END_BENCH -->/ { print; in_bench = 0; next }
            !in_bench { print }
          ' README.md > README.md.tmp && mv README.md.tmp README.md

      - name: Commit benchmark results to README
        run: |
          git config --global user.name "github-actions[bot]"
          git config --global user.email "github-actions@users.noreply.github.com"
          git diff --quiet || (git add README.md && git commit -m "Auto-update README with example code" && git push)

  deploy:
    needs: bench
    runs-on: ubuntu-latest
    permissions:
      pages: write
      id-token: write
    environment:
      name: github-pages
      url: ${{ steps.deployment.outputs.page_url }}
    steps:
      - name: Deploy to GitHub Pages
        id: deployment
        uses: actions/deploy-pages@v4
