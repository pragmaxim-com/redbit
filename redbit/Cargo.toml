[package]
name = "redbit"
version = "0.1.0"
edition = "2021"

[lib]
name = "redbit"

[dependencies]
macros = { path = "../macros" }
redb = "2.4.0"