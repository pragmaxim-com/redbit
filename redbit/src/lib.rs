pub use macros::Redbit;
pub use redb::ReadableTable;

#[derive(Debug)]
pub enum DbEngineError {
    DbError(String),
}

impl std::fmt::Display for DbEngineError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            DbEngineError::DbError(msg) => write!(f, "Database error: {}", msg),
        }
    }
}

impl std::error::Error for DbEngineError {}

impl From<redb::Error> for DbEngineError {
    fn from(e: redb::Error) -> Self {
        DbEngineError::DbError(e.to_string())
    }
}
impl From<redb::TransactionError> for DbEngineError {
    fn from(e: redb::TransactionError) -> Self {
        DbEngineError::DbError(e.to_string())
    }
}
impl From<redb::StorageError> for DbEngineError {
    fn from(e: redb::StorageError) -> Self {
        DbEngineError::DbError(e.to_string())
    }
}
impl From<redb::TableError> for DbEngineError {
    fn from(e: redb::TableError) -> Self {
        DbEngineError::DbError(e.to_string())
    }
}
impl From<redb::CommitError> for DbEngineError {
    fn from(e: redb::CommitError) -> Self {
        DbEngineError::DbError(e.to_string())
    }
}