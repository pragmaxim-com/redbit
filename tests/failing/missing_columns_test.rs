use redbit::*;

#[derive(Redbit)]
struct MissingColumnsStruct {
    #[pk]
    id: u32,
}

fn main() {
    // If it compiles successfully, we're good.
}
