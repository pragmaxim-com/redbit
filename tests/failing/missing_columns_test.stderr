error: No #[column(...)] fields found. You must have at least one field with #[column].
 --> tests/failing/missing_columns_test.rs:4:1
  |
4 | struct MissingColumnsStruct {
  | ^^^^^^
